tests/Utilities/testframework.f90

```diff
@@ -0,0 +1,143 @@
+
+CONTAINS
+
+  !> Initialise contents of test framework. Must be done before first use.
+  SUBROUTINE INITIALISE_TESTS()
+    current_test_name = ""
+    test_err = 0
+  END SUBROUTINE INITIALISE_TESTS
+
+  !> Clean up anything used in test framework and return result. Call after last test.
+  SUBROUTINE FINALISE_TESTS(err)
+    INTEGER(CMISSIntg), INTENT(OUT) :: err !< The test result, non-zero if failed.
+    err = test_err
+  END SUBROUTINE FINALISE_TESTS
+
+  !> Begin test of the given name. Currently doesn't handle nested tests. Must pair with END_TEST
+  SUBROUTINE BEGIN_TEST(test_name)
+    CHARACTER(LEN=*), INTENT(IN) :: test_name !< The name of the test.
+
+    IF (LEN(current_test_name) > 0) THEN
+      WRITE(unit = error_unit, fmt = "(a)") "Called BEGIN_TEST when test in progress. Cannot nest tests. Aborting."
+      STOP 1
+    ENDIF
+    WRITE(unit = output_unit, fmt = "(a)") "----------------------------------------"
+    WRITE(unit = output_unit, fmt = "(a,a)") "Begin test: ", test_name
+    current_test_name = test_name
+  END SUBROUTINE BEGIN_TEST
+
+  !> End the current test. Test name is retained from prior call to BEGIN_TEST.
+  SUBROUTINE END_TEST()
+    IF (LEN(current_test_name) == 0) THEN
+      WRITE(unit = error_unit, fmt = "(a)") "Called END_TEST without call to BEGIN_TEST. Aborting."
+      STOP 1
+    ENDIF
+    WRITE(unit = output_unit, fmt = "(a,a)") "End test: ", current_test_name
+    current_test_name = ""
+  END SUBROUTINE END_TEST
+
+  !> Record that an error has occurred.
+  SUBROUTINE set_err()
+    test_err = 1
+  END SUBROUTINE
+
+  !> Generate a unique user number for an object, by incrementing an integer module variable.
+  FUNCTION AUTO_USER_NUMBER() RESULT(user_number)
+    INTEGER(CMISSIntg) :: user_number
+    ! static variables
+    INTEGER(CMISSIntg), SAVE :: next_user_number = 1
+
+    user_number = next_user_number
+    next_user_number = next_user_number + 1
+  END FUNCTION AUTO_USER_NUMBER
+
+  !> Check actual value is equal to expect, integer variant. Reports and records any error and continues.
+  SUBROUTINE EXPECT_EQ_INTG(description, expected, actual)
+    CHARACTER(LEN=*), INTENT(IN) :: description !< Description of quantity being compared.
+    INTEGER(CMISSIntg), INTENT(IN) :: expected !< Expected value.
+    INTEGER(CMISSIntg), INTENT(IN) :: actual !< Actual value.
+
+    IF (actual /= expected) THEN
+      WRITE(unit = output_unit, fmt = "(a,a,a,i0,a,i0,a)") "The value of ", description, " (", actual, &
+        & ") is not near expected (", expected, ")"
+      CALL set_err()
+    ENDIF
+  END SUBROUTINE EXPECT_EQ_INTG
+
+  !> Check actual value is within tolerance of expected. Reports and records any error and continues.
+  SUBROUTINE EXPECT_NEAR_DP(description, expected, actual, tolerance)
+    CHARACTER(LEN=*), INTENT(IN) :: description !< Description of quantity being compared.
+    REAL(DP), INTENT(IN) :: expected !< Expected value.
```

Comment:
It is for the future as I'm just adding the RP stuff (next commit). RP is set to either SP or DP when the library is compiled. The program that calls the library should use CMISSRP always. I've added a cmfe_WorkingRealPrecisionGet(workingPrecision,err) subroutine that will return either CMFE_SINGLE_REAL_TYPE or CMFE_DOUBLE_REAL_TYPE but this is not pulled in yet. The precision() or kind() routines will take a fortran variable and tell you the precision and or kind so this is not really what you want.